applicationinsights-react-native/src/ReactNativeManualDevicePlugin.ts

```diff
@@ -0,0 +1,319 @@
+    AnalyticsPluginIdentifier, IAppInsights, IDevice, IExceptionTelemetry, eSeverityLevel
+} from "@microsoft/applicationinsights-common";
+import {
+    BaseTelemetryPlugin, IAppInsightsCore, IConfigDefaults, IPlugin, IProcessTelemetryContext, IProcessTelemetryUnloadContext, ITelemetryItem,
+    ITelemetryPlugin, ITelemetryUnloadState, _eInternalMessageId, _throwInternal, _warnToConsole, arrForEach, dumpObj, eLoggingSeverity,
+    getExceptionName, onConfigChange
+} from "@microsoft/applicationinsights-core-js";
+import { getGlobal, strShimUndefined } from "@microsoft/applicationinsights-shims";
+import { INativeDevice, IReactNativePluginConfig } from "./Interfaces";
+import { isPromiseLike, isString, ITimerHandler, objDeepFreeze, scheduleTimeout } from "@nevware21/ts-utils";
+import { IDeviceInfoModule } from "./Interfaces/IDeviceInfoModule";
+
+declare var global: Window;
+const defaultReactNativePluginConfig: IConfigDefaults<IReactNativePluginConfig> = objDeepFreeze({
+    disableDeviceCollection: false,
+    disableExceptionCollection: false,
+    uniqueIdPromiseTimeout: 5000
+});
+
+export class ReactNativeManualDevicePlugin extends BaseTelemetryPlugin {
+
+    identifier: string = "AppInsightsReactNativePlugin";
+    priority: number = 140;
+    _nextPlugin?: ITelemetryPlugin;
+
+    private _setExceptionHandler: () => void;
+    private _collectDeviceInfo: () => boolean;
+
+    constructor(config?: IReactNativePluginConfig) {
+        super();
+
+        // Automatic defaults, don't set values here only set in  _initDefaults()
+        let _device: INativeDevice;
+        let _config: IReactNativePluginConfig;
+        let _analyticsPlugin: IAppInsights;
+        let _defaultHandler;
+        let _waitingForId: boolean;
+        let _waitingTimer: ITimerHandler;
+        let _waitingItems: { item: ITelemetryItem, itemCtx?: IProcessTelemetryContext }[] = null;
+        let _deviceInfoModule: IDeviceInfoModule;
+       
+        let _deviceInfoNeedsUpdate:boolean;
+        let exceptionHandlerSet:boolean;
+    
+        dynamicProto(ReactNativeManualDevicePlugin, this, (_self, _base) => {
+            _initDefaults();
+
+            _self.initialize = (
+                config?: IReactNativePluginConfig | object, // need `| object` to coerce to interface
+                core?: IAppInsightsCore,
+                extensions?: IPlugin[]
+            ) => {
+                let identifier = this.identifier;
+                if (!_self.isInitialized()) {
+                    _base.initialize(config, core, extensions);
+
+                    _self._addHook(onConfigChange(config, (details) => {
+                        let ctx = _self._getTelCtx();
+                        _config = ctx.getExtCfg<IReactNativePluginConfig>(identifier, defaultReactNativePluginConfig);
+
+                        if (!_config.disableDeviceCollection && _deviceInfoNeedsUpdate) {
+                            _deviceInfoNeedsUpdate = !_self._collectDeviceInfo();
```

Comment:
if we fail to collect device info at first try,  the _deviceInfoNeedsUpdate  should be set to true so that when user give the device later, we could  collect it again.